nexus/tests/integration_tests/disks.rs

```diff
@@ -822,6 +829,126 @@ async fn test_disk_reject_total_size_not_divisible_by_min_disk_size(
     );
 }
 
+async fn create_instance_with_disk(client: &ClientTestContext) {
+    create_instance_with(
+        &client,
+        ORG_NAME,
+        PROJECT_NAME,
+        INSTANCE_NAME,
+        &params::InstanceNetworkInterfaceAttachment::Default,
+        vec![params::InstanceDiskAttachment::Attach(
+            params::InstanceDiskAttach { name: DISK_NAME.parse().unwrap() },
+        )],
+    )
+    .await;
+}
+
+const ALL_METRICS: [&'static str; 6] =
+    ["activated", "read", "write", "read_bytes", "write_bytes", "flush"];
+
+#[nexus_test]
+async fn test_disk_metrics(cptestctx: &ControlPlaneTestContext) {
+    let client = &cptestctx.external_client;
+    DiskTest::new(&cptestctx).await;
+    create_org_and_project(client).await;
+    create_disk(&client, ORG_NAME, PROJECT_NAME, DISK_NAME).await;
+
+    // Whenever we grab this URL, get the last few seconds of metrics.
+    let metric_url = |metric_type: &str| {
+        let disk_url = format!("{}/{}", get_disks_url(), DISK_NAME);
+        format!(
+            "{disk_url}/metrics/{metric_type}?start_time={:?}&end_time={:?}",
+            Utc::now() - chrono::Duration::seconds(5),
+            Utc::now(),
+        )
+    };
+
+    // Try accessing metrics before we attach the disk to an instance.
+    //
+    // Observe that no metrics exist yet; no "upstairs" should have been
+    // instantiated on a sled.
+    let measurements: ResultsPage<Measurement> =
+        objects_list_page_authz(client, &metric_url("read")).await;
+    assert!(measurements.items.is_empty());
+
+    // Create an instance, attach the disk to it.
+    create_instance_with_disk(client).await;
+
+    // NOTE: This is a little gnarly, but we sleep long enough to allow some
+    // metrics to be populated. Noted that this relies on the producer server
+    // within the simulated disk using an interval smaller than a second.
+    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
```

Comment:
I hit this a bunch writing the original integration tests of the timeseries database. I generally had small retry-loops, using `retry_notify` or something similar. Not a big deal now, but probably better to move that way in the long-term.

Reply:
Good call, this is worthwhile doing. I added a `query_for_metrics_until_they_exist` helper.